### ISSUE TITLE
S3 with chunking never sends Preflight Access-Control-Allow-Oirigin as it never passes request to backend

### ISSUE THREAD

**Issue by gullevek** [FineUploader/fine-uploader#2031]
## Type of issue
- [X] Bug report
- [ ] Feature request

## Uploader type
- [ ] Traditional
- [X] S3
- [ ] Azure

### Note: Support requests cannot be accepted due to lack of time.

Running S3 type and PHP backend (from the example) with chunking and cors enabled. If the file is smaller than the chunking size or chunking is disabled the upload works. But once the chunking is enabled or chunking size is reached the upload fails for each PUT as the initial OPTIONS request response misses the "Access-Control-Allow-Origin" has this request is never sent to the backend.

This bug also exist on the demo page of fine uploader (https://fineuploader.com/demos#amazon-demo)

Summary:
OPTIONS request is never passed to backend

#### Fine Uploader version
5.16.2

#### Browsers where the bug is reproducible
Firefox 61.0.1
Chrome 67.0.3396.99

#### Operating systems where the bug is reproducible
macOS 10.12.6

#### Exact steps required to reproduce the issue
For example:
1. got to the demo page: https://fineuploader.com/demos#amazon-demo
2. Upload a file of about 12MB
3. Look in the debug logs for the error:
DEMO:
Failed to load https://s3-demo.fineuploader.com/s3demo-thumbnails-cors.php: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'https://fineuploader.com' is therefore not allowed access. The response had HTTP status code 404.

My System:
Failed to load https://BUCKETNAME.s3.amazonaws.com/ac200326-520a-4f12-a49a-ff2cf3ed1aa3.bz2?partNumber=3&uploadId=RETRACTED-UPLOAD-ID: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://MYDOMAIN.COM' is therefore not allowed access. The response had HTTP status code 403.

![fu - console log](https://user-images.githubusercontent.com/855861/43182130-227d8d60-901b-11e8-9030-f5b8e3a844a5.png)

#### All relevant Fine Uploader-related code that you have written
The Javascript

```
<script language="JavaScript">
var s3Uploader = new qq.s3.FineUploader({
	debug: true,
	element: $('fine-uploader-s3-1'),
	template: 'qq-template-s3',
	multiple: false, // no multiple uploads allowed
	request: {
		endpoint: "https://MYBUCKET.s3.amazonaws.com",
		accessKey: "MYS3ACCESSKEY", // this key should be set from script on load
		params: {
			category: "TEST-CATEGORY"
		}
	},
	// local calls
	signature: {
		endpoint: "/fineUploader/s3/endpoint-cors.php"
	},
	uploadSuccess: {
		endpoint: "/fineUploader/s3/endpoint-cors.php?success",
	},
	iframeSupport: {
		localBlankPagePath: "/fineUploader/s3/success.html"
	},
	retry: {
		enabledAuto: true,
		showButton: true
	},
	cors: {
		expected: true
	},
	chunking: {
		enabled: true,
		concurrent: {
			enabled: true
		}
	},
	resume: {
		enabled: true
	},
	deleteFile: {
		enabled: true,
		method: "POST",
		endpoint: "/fineUploader/s3/endpoint-cors.php"
	},
	callbacks: {
		onComplete: function(id, name, response) {
			console.log('Completed for ID: %s, with name: %s', id, name);
		}
	}
});
</script>

```

S3 CORS config:
```
<?xml version="1.0" encoding="UTF-8"?>
<CORSConfiguration xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
<CORSRule>
    <AllowedOrigin>*</AllowedOrigin>
    <AllowedMethod>POST</AllowedMethod>
    <AllowedMethod>GET</AllowedMethod>
    <AllowedMethod>DELETE</AllowedMethod>
    <MaxAgeSeconds>3000</MaxAgeSeconds>
    <ExposeHeader>ETag</ExposeHeader>
    <AllowedHeader>*</AllowedHeader>
</CORSRule>
</CORSConfiguration>

```

#### Detailed explanation of the problem

_OPTIONS REQUEST:_

> Request URL: https://MYBUCKET.s3.amazonaws.com/07759238-c976-44e6-a17d-4b77211f483a.bz2?uploads
> Request method: OPTIONS
> Remote address: 52.216.131.163:443
> 

Request headers:

> Host: MYBUCKET.s3.amazonaws.com
> User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:61.0) Gecko/20100101 Firefox/61.0
> Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
> Accept-Language: en-US,en;q=0.5
> Accept-Encoding: gzip, deflate, br
> Access-Control-Request-Method: POST
> Access-Control-Request-Headers: authorization,content-type,x-amz-acl,x-amz-date,x-amz-meta-category,x-amz-meta-qqfilename
> Origin: http://MYHOSTNAME.com
> Connection: keep-alive

Response headers:

> HTTP/1.1 200 OK
> x-amz-id-2: <AUTH ID>
> x-amz-request-id: <REQUEST ID>
> Date: Wed, 25 Jul 2018 05:37:19 GMT
> Access-Control-Allow-Origin: *
> Access-Control-Allow-Methods: POST, GET, DELETE
> Access-Control-Allow-Headers: authorization, content-type, x-amz-acl, x-amz-date, x-amz-meta-category, x-amz-meta-qqfilename
> Access-Control-Expose-Headers: ETag
> Access-Control-Max-Age: 3000
> Vary: Origin, Access-Control-Request-Headers, Access-Control-Request-Method
> Content-Length: 0
> Server: AmazonS3

![fu - 1 - options](https://user-images.githubusercontent.com/855861/43182141-2f024e18-901b-11e8-8f7b-5c4f028249ac.png)

_Follow up POST_

> Request URL: https://MYBUCKET.s3.amazonaws.com/07759238-c976-44e6-a17d-4b77211f483a.bz2?uploads
> Request method: POST
> Remote address: 52.216.131.163:443

Request Headers:

> Host: MYBUCKET.s3.amazonaws.com
> User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.12; rv:61.0) Gecko/20100101 Firefox/61.0
> Accept: */*
> Accept-Language: en-US,en;q=0.5
> Accept-Encoding: gzip, deflate, br
> Referer: http://MYHOST.COM/fineUploader/fineuploader.s3.front.php
> x-amz-acl: private
> x-amz-meta-qqfilename: original-file-name.tar.bz2
> x-amz-meta-category: CATEGORY-TEST
> x-amz-date: Wed, 25 Jul 2018 05:37:17 GMT
> Content-Type: application/x-bzip2
> Authorization: AWS <AUTH KEY>
> Origin: http://MYHOSTNAME.COM
> Connection: keep-alive
> Content-Length: 0

Response headers:

> HTTP/1.1 200 OK
> x-amz-id-2: 19MoBhRrJrlagXJPVTTrZX7CN9iS2qfT6fmYDVT31UiAYQWp6uf4/LzoTDwmafBnYkN4JtAmqE0=
> x-amz-request-id: 4CE1153A572D0731
> Date: Wed, 25 Jul 2018 05:37:19 GMT
> Access-Control-Allow-Origin: *
> Access-Control-Allow-Methods: POST, GET, DELETE
> Access-Control-Expose-Headers: ETag
> Access-Control-Max-Age: 3000
> Vary: Origin, Access-Control-Request-Headers, Access-Control-Request-Method
> Transfer-Encoding: chunked
> Server: AmazonS3

![fu - 2 - post](https://user-images.githubusercontent.com/855861/43182151-3814ec36-901b-11e8-8ec4-aa52cf662385.png)

The following POSTs that do the PUT are sent through the PHP backend script and have the header set

Response headers:

> HTTP/1.1 200 OK
> Date: Wed, 25 Jul 2018 05:37:18 GMT
> Server: Apache/2.4.29 (Unix) PHP/7.2.8
> X-Powered-By: PHP/7.2.8
> **Access-Control-Allow-Origin: http://MYHOST.COM** <-- HERE IT IS
> Content-Length: 44
> Keep-Alive: timeout=5, max=99
> Connection: Keep-Alive
> Content-Type: application/json

![fu - 3 - post - put](https://user-images.githubusercontent.com/855861/43182159-3f19c998-901b-11e8-87c4-75a7c06cb635.png)

I also added debug into the s3 PHP backend script as provided from here (https://github.com/FineUploader/php-s3-server/blob/master/endpoint-cors.php)

> {5b580f9f6893d} [2018-07-25 14:07:50.42887100] <S3 ENDPOINT> **** File: endpoint-cors.php, Method: POST, POST: Array
> {5b580fa088c30} [2018-07-25 14:07:50.56066600] <S3 ENDPOINT> **** File: endpoint-cors.php, Method: POST, POST: Array
> {5b580fa089e1c} [2018-07-25 14:07:50.56532200] <S3 ENDPOINT> **** File: endpoint-cors.php, Method: POST, POST: Array
> {5b580fa0936be} [2018-07-25 14:07:50.60435600] <S3 ENDPOINT> **** File: endpoint-cors.php, Method: POST, POST: Array

Four request for four split signatures, but no OPTIONS request for the initial REST part to S3

And I cannot see ANY method request as OPTIONS as to which it should send out the headers. I also cannot find where in the JS code this request should be sent. In all the documentations (eg the great blog post about S3 implementation) is anything special written for the config to be changed for this.

**Comment by rnicholus** [comment]
Fine Uploader doesn’t send preflight requests. That’s not how CORS works at all. The browser and only the browser controls the preflight request. The demo site issue you referenced has nothing to do with CORS. Notice the response is a 404, since I stopped paying the AWS bills long ago. And on your server, again, the error has nothing to do with CORS. Your server is returning a 403, so you’ll want to look closer at that.

**Comment by gullevek** [comment]
Thank you, I just thought OPTIONS requests are forwraded to the defined endpoint somewhere, but ti seems they end somewhere. Perhaps I can find some target in Chrome or Firefox on what level or where I can actually response to them.

And 403 error means non-authorized because the missing pre-flight Access-Control-Allow-Origin header

**Comment by rnicholus** [comment]
That header is not sent to you, you are expected to send it in the response.

**Comment by gullevek** [comment]
But in what response? I never see any response to my server that could trigger those headers. That's why I am so confused about this error.

JS-> PUT -> triggers OPTIONS from browser to S3 -> S3 sends back "I can do this"
so where do I then trigger this response if my sever never sees that? Or do I have to put that in JS in the web browser? (which would make no sense)

**Comment by gullevek** [comment]
I found the problem. It had NOTHING to with your script nor with my server settings. I made a mistake in the CORS settings on the s3 server.

**I had a "GET" allow instead of a "PUT" allow**

Sadly no browser gives you correct error reports about this at all. What a pitty.
I only found out about this when I looked in the Chrome deny logs where it shows Access control request method: PUT. Firefox just never shows them, just the initial POST one.

Jesus :(
